apps/site/assets/ts/stop/components/DepartureTimes.tsx

```diff
@@ -1,74 +1,299 @@
 import renderFa from "../../helpers/render-fa";
-import { formatDepartureTime } from "../../helpers/date";
+import { formatToBostonTime } from "../../helpers/date";
+import { find, groupBy, slice, some } from "lodash";
+import {
+  differenceInSeconds,
+  isSameDay,
+  secondsInDay,
+  secondsInHour,
+  secondsInMinute
+} from "date-fns";
+import realtimeIcon from "../../../static/images/icon-realtime-tracking.svg";
+import SVGIcon from "../../helpers/render-svg";
+import { ScheduleWithTimestamp } from "../../models/schedules";
+import { PredictionWithTimestamp } from "../../models/perdictions";
+import {
+  departureInfoToTime,
+  mergeIntoDepartureInfo
+} from "../../helpers/departureInfo";
+import { DepartureInfo } from "../../models/departureInfo";
 
-interface DepartureTimesProps {
-  route: Route;
-  stop: Stop;
-  directionId: DirectionId;
+// This interface is used to tell the front end
+// how to display the ScheduleInfoModel data
+interface DisplayTimeConfig {
+  isPrediction?: boolean;
+  displayString: string;
+  trackName?: string | null;
+  isTomorrow?: boolean;
+  isBolded?: boolean;
+  isStikethrough?: boolean;
 }
 
+const getNextUnCancelledDeparture = (
+  departureInfos: DepartureInfo[]
+): DepartureInfo | undefined => {
+  return find(departureInfos, (di: DepartureInfo) => !di.isCancelled);
+};
+
+// This function returns time info in a consisten format.
+// Returns the time from the predictions if they exist, or the schdule
+const getNextTwoTimes = (
+  departureInfos: DepartureInfo[]
+): [boolean, DepartureInfo | undefined, DepartureInfo | undefined] => {
+  const departure1 = departureInfos[0];
+  const departure2 = getNextUnCancelledDeparture(slice(departureInfos, 1));
+
+  // TODO is this needed to calculate
+  const areAnyPredictions = some(
+    [departure1, departure2],
+    (d: DepartureInfo | undefined) => d && d.prediction
+  );
+
+  return [areAnyPredictions, departure1, departure2];
+};
+
+const infoToDisplayTime = (
+  time1: DepartureInfo | undefined,
+  time2: DepartureInfo | undefined,
+  isPrediction: boolean,
+  targetDate: Date = new Date()
+): DisplayTimeConfig[] => {
+  const defaultState = [{ displayString: "Error retrieving data" }];
+  // If there is not input time1 then a schedule or prediction could not be found
+  if (!time1) {
+    return defaultState;
+  }
+
+  const departure1Time = departureInfoToTime(time1);
+  const formatOverride = "h:mm aa";
+
+  if (time2 && (time1.isCancelled || time1.isDelayed)) {
+    const departure2Time = departureInfoToTime(time2);
+    // State 7
+    // State 8
+    // If trip1 is cancelled/delayed, then trip2 should not be cancelled/delayed
+    // Display trip2 in the first time spot (and its track info in the second)
+    return [
+      {
+        displayString: `${formatToBostonTime(departure2Time, formatOverride)}`,
+        isBolded: true,
+        // Only predictions can be deleayed or cancelled
+        isPrediction: true
```

Comment:
This was originally hard coded to true, because I thought (wrongly) we were displaying the indicator if any time was a prediction.  This has been changed for cancellations to check if the new time is a prediction.

As for delays this is always true.  That is because the way we test for delays is comparing the schedule and the prediction.  Is the prediction is more than 60 seconds after the schedule, it is delayed.  The prediction time is shown.  That makes it always a prediction.

`isDelayed` and `isCancelled` are getting broken out into their own branch.